src/Doctrine/Common/State/PersistProcessor.php

```diff
@@ -52,10 +52,34 @@ public function process(mixed $data, Operation $operation, array $uriVariables =
         // https://github.com/doctrine/orm/issues/8461#issuecomment-1250233555
         if ($operation instanceof HttpOperation && HttpOperation::METHOD_PUT === $operation->getMethod() && ($operation->getExtraProperties()['standard_put'] ?? false)) {
             \assert(method_exists($manager, 'getReference'));
-            // TODO: the call to getReference is most likely to fail with complex identifiers
             $newData = $data;
             if (isset($context['previous_data'])) {
-                $newData = 1 === \count($uriVariables) ? $manager->getReference($class, current($uriVariables)) : clone $context['previous_data'];
+                // the correct entity need to be fetched, we need to get the ORM ids from previous data
+                $reflectionProperties = $this->getReflectionProperties($context['previous_data']);
+                $ormIds = [];
+                \assert(method_exists($manager, 'getClassMetadata'));
+                foreach ($manager->getClassMetadata($class)->getIdentifier() as $identifier) {
+                    if (\array_key_exists($identifier, $reflectionProperties)) {
+                        $value = $reflectionProperties[$identifier]->getValue($context['previous_data']);
+                    } else {
+                        // This id is not from current class, let's move up in class hierarchy
+                        $parent = (new \ReflectionObject($data))->getParentClass();
+                        while (!isset($value) && false !== $parent) {
+                            foreach ($parent->getProperties() as $property) {
+                                if ($property->name === $identifier) {
+                                    $value = $property->getValue($context['previous_data']);
```

Comment:
this looks a lot like https://github.com/api-platform/core/blob/9b4b58ca0113a2645b58a116fe9e4bf200df8aa3/src/Api/IdentifiersExtractor.php#L96 we should already have extracted the identifiers and they should be available inside `uriVariables` at this point.

Reply:
This does look a lot like what IdentifierExtractor does and likely needs to be replaced, but i disagree on the last part : there's no reason for these values to appear in `uriVariables` in this case, or i'm missing something.